[Merged by Bors] - feat(Algebra/MonoidAlgebra): complete the map API

YaelDillies opened the pull request:
Add congruence isomorphisms in both the ring and monoid arguments. Move `mapRangeRingHom` from `Algebra.MonoidAlgebra.Lift` to `Algebra.MonoidAlgebra.MapDomain` and make its proofs additivisable (by not using `lift`). Make `mapDomainRingHom` take a `MonoidHom` rather than `MonoidHomClass`.

---
- [x] depends on: #32486

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -88,6 +88,60 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
+namespace RingEquiv
+variable [Semiring R] [Semiring S] [Monoid M]
+
+/-- Isomorphic rings have isomorphic monoid algebras. -/
```
```suggestion
/-- Isomorphic rings have isomorphic additive monoid algebras. -/
```

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -88,6 +88,60 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
+namespace MulEquiv
+variable [Semiring R] [Monoid M] [Monoid N]
+
+/-- Isomorphic monoids have isomorphic monoid algebras. -/
```
```suggestion
/-- Isomorphic monoids have isomorphic additive monoid algebras. -/
```

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -88,6 +88,60 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
+
+/-- Isomorphic monoids have isomorphic monoid algebras. -/
+@[to_additive (dont_translate := R) (attr := simps apply symm_apply)
+/-- Isomorphic monoids have isomorphic monoid algebras. -/]
```
```suggestion
/-- Isomorphic monoids have isomorphic additive monoid algebras. -/]
```

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -88,6 +92,104 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
+/-- Isomorphic rings have isomorphic monoid algebras. -/
+@[to_additive (dont_translate := R S) (attr := simps! apply symm_apply)
+/-- Isomorphic rings have isomorphic additive monoid algebras. -/]
+def monoidAlgebraCongrLeft (e : R ≃+* S) : MonoidAlgebra R M ≃+* MonoidAlgebra S M :=
```
This name doesn't match the existing `MonoidAlgebra.mapRangeAlgEquiv`

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -88,6 +92,104 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
+/-- Isomorphic monoids have isomorphic monoid algebras. -/
+@[to_additive (dont_translate := R) (attr := simps! apply symm_apply)
+/-- Isomorphic monoids have isomorphic additive monoid algebras. -/]
+def monoidAlgebraCongrRight (e : M ≃* N) : MonoidAlgebra R M ≃+* MonoidAlgebra R N :=
```
This is a weaker version of `MonoidAlgebra.domCongr`; do we need both?

YaelDillies replied:
`MonoidAlgebra.domCongr` is also defined too late to be used in the places where I want to use without increasing imports

YaelDillies replied:
Not without removing some `assert_not_exists`s. But in other places of the library we tend to keep the `RingEquiv` versions even when they are not more general than the `AlgEquiv` ones, so I claim this is okay.

YaelDillies committed:
feat(Algebra/MonoidAlgebra): complete the map API
Add congruence isomorphisms in both the ring and monoid arguments. Move `mapRangeRingHom` from `Algebra.MonoidAlgebra.Lift` to `Algebra.MonoidAlgebra.MapDomain` and make its proofs additivisable (by not using `lift`). Make `mapDomainRingHom` take a `MonoidHom` rather than `MonoidHomClass`.

YaelDillies committed:
use it in `domCongr`

YaelDillies committed:
fix
